### PR TITLE
Initial release of ``django-pglock``

### PR DESCRIPTION
**Pull request by wesleykendall**
``django-pglock`` performs advisory locks, table locks, and helps manage blocking locks. Here's some of the functionality at a glance:

* ``pglock.advisory`` for application-level locking, for example, ensuring that tasks don't overlap.
* ``pglock.model`` for locking an entire model.
* ``pglock.timeout`` for dynamically setting the timeout to acquire a lock.
* ``pglock.bypass`` to kill blocking locks for critical code, such as migrations.
* The ``PGLock`` and ``BlockedPGLock`` models for querying active and blocked locks.
* The ``pglock`` management command that wraps the models and provides other utilities.

Type: api-break

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,11 +1,138 @@
...
+
+``pglock.model`` can take a lock on an entire model during a transaction. For example:
+
+.. code-block:: python
+
+    from django.db import transaction
+    import pglock
+
+    with transaction.atomic():
+        pglock.model("auth.User")
```
Comment (PaulGilmartin) [inline README.rst:63]: Maybe it's explained later, but I'm still not sure where one would actually use this in practice?

File: README.rst [hunk 2]
```diff
@@ -1,11 +1,138 @@
...
+
+        # Any operations on auth.User will be exclusive here. Even read access
+        # for other transactions is blocked
```
Comment (PaulGilmartin) [inline README.rst:66]: Maybe it's explained later, but I'm still not sure where one would actually use this in practice?

File: README.rst [hunk 3]
```diff
@@ -1,11 +1,138 @@
...
+    import pglock
+
+    with pglock.advisory("my_lock_id"):
+        # This code blocks until the "my_lock_id" lock is available
+
+
+Above our code will block until the lock is available, meaning
+no instances of the function will run simultaneously.
+
+Use ``wait=False`` or ``timeout=pglock.timedelta(seconds=...)`` to avoid waiting:
```
Comment (PaulGilmartin) [inline README.rst]: Some inconsistency between the use of timeout=pglock.timedelta(seconds=...) here and @pglock.timeout(seconds=1) below

File: README.rst [hunk 4]
```diff
@@ -1,11 +1,138 @@
...
+Use ``pglock.timeout`` to dynamically set `Postgres's lock_timeout runtime
+setting <https://www.postgresql.org/docs/current/runtime-config-client.html>`__:
+
+.. code-block:: python
+
+    import pglock
+
+    @pglock.timeout(seconds=1)
+    def do_stuff():
+        # This function will thrown an exception if any code takes longer than
```
Comment (PaulGilmartin) [inline README.rst]: Typo "thrown"

File: docs/advisory.rst [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+Above we've ensured that we wait at most two seconds for the lock to be acquired.
+`pglock.timedelta` is simply an alias for Python's ``timedelta`` object for convenience.
+
+.. note::
+
+    You might be asking, why not just use ``timeout=pglock.timedelta(seconds=0)`` instead
+    of ``wait=False``? These arguments are separate because they implement different
+    strategies under the hood. ``wait=False`` uses a different Postgres function
+    to acquire the lock. It's recommended to use this when waiting isn't required.
```
Comment (PaulGilmartin) [inline docs/advisory.rst]: Rather than revealing this internal detail to the user, would it not be better to just implicitly use the current `wait=False` behaviour under the hood if someone passes in ``timeout=pglock.timedelta(seconds=0)``? Then we can actually just remove the `wait` argument if we want. It seems perhaps confusing to the end user that two different arguments can have the same semantic meaning to them but do completely different things under the hood.

File: pglock/core.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+        if not self.lock_id:
+            module = inspect.getmodule(func)
+            self.lock_id = f"{module.__name__}.{func.__name__}"
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            with self._recreate_cm():
+                if self.acquired or self.side_effect != SKIP:
+                    return func(*args, **kwargs)
```
Comment (PaulGilmartin) [inline pglock/core.py:234]: In the docs it's mentioned that "When used as a decorator, the return value of pglock.advisory is not available, making it impossible to know if the lock was acquired".

If we did want to make it available, couldn't we just set the `self.acquired` value as an instance attribute on `func` here?

```
        @functools.wraps(func)
        def inner(*args, **kwargs):
            with self._recreate_cm():
                   func.lock_acquired = True
                   return func(*args, **kwargs)
```

Then one can make use of this within the scope of func:

```
@pglock.advisory
def my_locked_func():
      if my_locked_func.lock_acquired: 
          raise OperationalError
```

File: pglock/core.py [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+            raise ValueError(
+                "Cannot use pglock.SKIP in a context manager."
+                " Use the @pglock.advisory decorator instead."
+            )
+
+        sql = f"SELECT {self.acquire}({self.int_lock_id})"
+
+        with connections[self.using].cursor() as cursor:
+            try:
+                with contextlib.ExitStack() as stack:
```
Comment (PaulGilmartin) [inline pglock/core.py:282]: I haven't ever seen ExitStack before. From what I now read, it seems we're using it here because we potentially do two nested context managers (lock timeout and atomic), and we want these to be cleaned up at the same time in the advisory.__exit__ method?

File: pglock/core.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+                    if self.timeout is not None:
+                        stack.enter_context(lock_timeout(self.timeout, using=self.using))
+
+                        if self.side_effect == RETURN and connections[self.using].in_atomic_block:
+                            # If returning True/False, create a savepoint so that
+                            # the transaction isn't in an errored state when returning.
```
Comment (PaulGilmartin) [inline pglock/core.py:288]: I don't fully understand what the savepoint is doing here

File: pglock/core.py [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+            module = inspect.getmodule(func)
+            self.lock_id = f"{module.__name__}.{func.__name__}"
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            with self._recreate_cm():
+                if self.acquired or self.side_effect != SKIP:
+                    return func(*args, **kwargs)
+
+        return inner
```
Comment (PaulGilmartin) [inline pglock/core.py:236]: It may also be worth mentioning here that one can reverse engineer an advisory lock from the lock_id.
When we take a lock by passing in this key, postgres creates an entry in pg_locks where
the key's high-order half in the classid column,  they key's low-order half in the objid column.

So, given the lock_id the user chose when setting the lock initially, we can check whether or not that lock_id actually acquired a lock using

```
def lower_and_upper_32_bits(value):
       lower = value & 0xffffffff
       upper = value >> 32
       return lower, upper

obj_id, class_id = lower_and_upper_32_bits(lock_id)

cursor.execute(f"""
            SELECT 1
            FROM pg_locks
            WHERE locktype = 'advisory'
              and mode = 'ExclusiveLock'
              and granted = 't'
              and classid = {class_id}
              and objid = {obj_id}
            """)
```

The advantage of this of course is that it allows us to check the acquired status when outside the scope of the acquiring action.

File: docs/bypass.rst [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+Here are the other arguments of `pglock.bypass` for more advanced use cases:
+
+* **timeout**: Set the lock timeout. This is equivalent to calling
+  ``with pglock.bypass(), pglock.timeout(...):``. This can be used
+  as an additional safety net in the case the background worker cannot
+  properly terminate blocking activity.
+* **retries**: Retry the side effect this many times.
+  No retries are performed by default.
+  For example, other blocking row-level locks
+  may arise immediately after other locks have been cleared. Using ``retries`` ensures another
```
Comment (PaulGilmartin) [inline docs/bypass.rst]: A little confused by this description. The scenario you describe where a lock can appear again immediately after terminating is a common one, but I can't see why we'd try to solve it by passing in a finite 'retries' value. For example, if I was the end user, passing in e.g. retries=2 would feel the same as not passing in a retry at all. If I can't be sure that the lock won't come back after a single terminate action, why can I be any more sure about it not coming back after 3 terminate actions?

It seems to be that this scenario would be better solve by something like a `continual` argument which just terminates in a loop?

File: docs/command.rst [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+3. **mode**: The mode of the lock, such as "ACCESS_EXCLUSIVE".
+4. **rel_kind**: The kind of relationship, such as "TABLE".
+5. **rel_name**: The name of the relationship locked, such as the table name.
+6. **activity__context**: Application context tracked by
+   `django-pgactivity <https://django-pgactivity.readthedocs.io>`__.
+7. **activity__query**: The SQL of the query.
+
+Output looks like the following::
+
+    76362 | 21:02:35 | ACCESS_EXCLUSIVE | TABLE | auth_user | None | lock table auth_user in access
```
Comment (PaulGilmartin) [inline docs/command.rst:27]: Does the output contain the column names?

File: docs/timeout.rst [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+`pglock.timeout` can be nested like so:
+
+.. code-block:: python
+
+    with pglock.timeout(seconds=2):
+        # Every statment here will have a lock timeout of 2 seconds
+
+        with pglock.timeout(seconds=5):
+            # Every statement will now have a lock timeout of 5 seconds
+
```
Comment (PaulGilmartin) [inline docs/timeout.rst:35]: In some contexts this may be slightly confusing. For example, suppose someone wrapped an entire view in a pglock.timeout(seconds=2) manager with the deliberate intention that nothing should ever wait for more than 2 seconds. Someone could then easily end up adding some pglock.timeout(seconds=10) wrapper around a piece of code which is hit by the view but is no were "near" the orginal wrapper (e.g. in a signal or something). To prevent that, we could have something like an `allow_nesting` kwarg, which would be passed in as False to pglock.timeout(seconds=2, allow_nested=False) in this scenario. Sort of like the reverse of Django atomic's 'durable' kwarg.

File: pglock/core.py [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+
+    Raises:
+        django.db.utils.OperationalError: When a timeout occurs
+    """
+    if interval is None:
+        interval = dt.timedelta(**timedelta_kwargs)
+
+    assert isinstance(interval, dt.timedelta)
+
+    if not hasattr(_timeout, "value"):
```
Comment (PaulGilmartin) [inline pglock/core.py:110]: This attribute name might clash with an application's of a local thread attribute.

File: pglock/core.py [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+            raise
+        else:
+            raise AssertionError
+
+
+def _bypass_bg_task(backend_pid, retries, side_effect):
+    BlockedPGLock = apps.get_model("pglock.BlockedPGLock")
+
+    for _ in range(retries + 1):
+        qset = BlockedPGLock.objects.pid(backend_pid)
```
Comment (PaulGilmartin) [inline pglock/core.py]: I need a bit of reminding on what exactly this is doing here.
Is backend_pid the id of the postgres connection being used by the function wrapped in `bypass`? If so, that means `qset = BlockedPGLock.objects.pid(backend_pid)` fetches everything _blocking_ that connection? So in the case of migrations it would only kill exactly those locks which are blocking the migration, and no more?

File: pglock/models.py [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+                                '([A-Z])','_\1',
+                                'g'
+                            )
+                        )
+                    ) AS mode,
+                    CASE pg_class.relkind
+                        WHEN 'r' THEN 'TABLE'::text
+                        WHEN 'i' THEN 'INDEX'::text
+                        WHEN 'S' THEN 'SEQUENCE'::text
+                        WHEN 't' THEN 'TOAST'::text
```
Comment (PaulGilmartin) [inline pglock/models.py:51]: what is 'TOAST'?

File: pglock/core.py (hunk 14 above)
Comment (PaulGilmartin) [reply pglock/core.py]: Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, _bypass_ is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

This also makes me think about the "reverse" behaviour of bypass. As I mentioned in the pgmigrate review, we often get the in prod scenario where the blocks coming from the migration _after_ it has acquired the lock are just as damaging as the blocks preventing it acquiring the lock. Could bypass have an option that once the lock is acquired, it then starts killing PIDs which _it itself_ is blocking?

Alternatively, if we knew these blocking queries came from certain functions, we could decorate them and tell them not to wait for certain other functions and/or locks. Rough ideas:

```
@pglock.terminate_on('migrate')     # could take a list of function paths  or django commands for example
def my_slow_update_func():
       ...
```
OR
```
@pglock.terminate_on(pglock.ACCESS_EXCLUSIVE) 
def my_slow_update_func():
       ...
```

The first one I'm not exactly sure how we'd do, but it's the most appealing. The second one above would already be good for avoiding clashes with migrations I think.

File: README.rst (hunk 3 above)
Comment (wesleykendall) [reply README.rst]: What do you mean? ``timeout`` is an argument to ``pglock.advisory`` that takes a timedelta object:

```
@pglock.advisory(timeout=pglock.timedelta(seconds=1))
```

``pglock.timeout`` takes either a timedelta:

```
@pglock.timeout(pglock.timedelta(seconds=1))
```

Or you can supply it with the same arguments that timedelta accepts:

```
@pglock.timeout(seconds=1, milliseconds=200)
```

File: docs/advisory.rst (hunk 5 above)
Comment (wesleykendall) [reply docs/advisory.rst]: I was originally against it because I thought it was lengthy to have to use the ``timeout`` argument and supply a timedelta every time. What I think I can do instead is allow ``timeout`` to take a number in seconds or a timedelta object.

I.e. `pglock.advisory(timeout=0)` will perform NOWAIT, and I can update `pglock.model` to have the same behavior

File: pglock/core.py (hunk 6 above)
Comment (wesleykendall) [reply pglock/core.py:234]: I have yet to see a decorator that works this way in practice. Do you have any examples of libraries that do it? Plus it would have to be a thread local or some other way of making it thread safe.

`mock.patch` supplies the return value as an argument to the decorated function, but that's a pretty special case for test functions that are not imported and called by outside code

I think this is something that can be added later if requested

File: README.rst (hunk 2 above)
Comment (wesleykendall) [reply README.rst:66]: I wish I could find the article I read that had a great example of using table locks to serialize code that was otherwise prone to race conditions. I imagine people mostly use it in this way or to proactively take out locks to avoid a deadlock.

I don't have great examples, but this is more to fill a hole in Django since there's no way to lock tables

File: pglock/core.py (hunk 7 above)
Comment (wesleykendall) [reply pglock/core.py:282]: Yea exactly, it's just IMO a cleaner way of doing conditional context managers. It's there to avoid code like this:

```
if self.side_effect == RETURN ...:
    with lock_timeout, atomic():
        yield
else:
    yield
```

File: pglock/core.py (hunk 8 above)
Comment (wesleykendall) [reply pglock/core.py:288]: Imagine that you do this as a user:

```
with transaction.atomic():
    with pglock.advisory(timeout=timedelta(seconds=1)) as acquired:
        # Do queries here...
        User.objects...
```

By default, a lock timeout will cause the transaction to be in an error state, meaning that the `User.objects` query will fail. A save point is made around the ``pg_advisory_lock..`` statement to prevent this from happening. If a timeout occurs, code can continue as normal.

There's a test case for this in the test suite that fails

File: pglock/core.py (hunk 9 above)
Comment (wesleykendall) [reply pglock/core.py:236]: What about this - Let's make a separate `pglock.advisory_id` function that returns the class_id/obj_id tuple and we can just use this internally inside of the `pglock.advisory` implementation. Anyone that wants to use it for their own needs can as well

File: pglock/core.py (hunk 6 above)
Comment (PaulGilmartin) [reply pglock/core.py:234]: Good point about the thread local nature of it, I had forgotten about that!
Another simple alternative is that we ask the user that the function they decorate accepts an `acquired` kwarg. But no need to change what we have.

File: docs/bypass.rst (hunk 10 above)
Comment (wesleykendall) [reply docs/bypass.rst]: Let's say you queue the following queries in order in separate sessions:

1. LOCK auth_user - Locks auth_user in access exclusive mode
2. BEGIN; SELECT * FROM auth_user FOR UPDATE;
3. BEGIN; SELECT * FROM auth_user FOR UPDATE;

When querying for the blocking locks of number 3, it only returns number 1. However, as soon as 1 is killed, 2 will begin and appear as blocking locks for number 3.

``retries`` was meant to alleviate this issue. Instead of waiting an entire interval to kill 3, it will just go ahead and kill it in the first interval before it pauses again.

I didn't make it work this way by default because I read that using `pg_blocking_pids` is an expensive operation, so I was hesitant to query it so much.

I'm not sure if this totally clears up the confusion, but what I'm solving here is ultimately an edge case. Even if it did happen in practice, the locks would still be terminated after some time.

For migration related locks like ACCESS EXCLUSIVE, this doesn't seem to be a problem because dynamic row lock queries will still be returned and can be killed in the first try.

TLDR - I will remove ``retries`` as an argument for now

File: docs/command.rst (hunk 11 above)
Comment (wesleykendall) [reply docs/command.rst:27]: `-e` for "expanded" mode shows the full SQL and the column names.

File: docs/timeout.rst (hunk 12 above)
Comment (wesleykendall) [reply docs/timeout.rst:35]: I'm hesitant to implement this feature since parts of the library depend on being able to dynamically set the lock timeout.

For example:

```
with pglock.timeout(seconds=1, allow_nested=False):

    # Now we can't do this...
    with pglock.advisory(timeout=2) as acquired:
        ...

    # We can't do this either...
    pglock.model(timeout=2)
```

I would rather wait to see if anyone requests this feature first

File: pglock/core.py (hunk 13 above)
Comment (wesleykendall) [reply pglock/core.py:110]: `_timeout` is a private global variable in the module. Not sure what you mean, but this is safe

File: pglock/core.py (hunk 14 above)
Comment (wesleykendall) [reply pglock/core.py]: > I need a bit of reminding on what exactly this is doing here.
Is backend_pid the id of the postgres connection being used by the function wrapped in bypass? If so, that means qset = BlockedPGLock.objects.pid(backend_pid) fetches everything blocking that connection? So in the case of migrations it would only kill exactly those locks which are blocking the migration, and no more?

Exactly! The `.pid()` method is first filtering the locks by their process ID. The `BlockedPGLock` model uses `pg_blocking_pids` to pull in the blocking PIDs for that process.

Note that underneath the hood I'm creating a CTE to query all locks. I made a special `.pid()` method to filter by the process ID (as opposed to just doing `BlockedPGLock.objects.filter(activity_id=pid)`) because this results in a much more efficient query for Postgres versions less than 12

Comment (wesleykendall) [reply pglock/core.py]: > Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, bypass is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

It's not the case because there could be multiple operations that get blocked later in code wrapped by `bypass`.

Let's say that a migration looks like this:

```
operations = [
    migrations.AlterField("model_one", field="field_one"),
    migrations.AlterField("model_two", field="field_two"),
    migrations.AlterField("model_three", field="field_three"),
]
```

The first operation may be blocked because of the exclusive access lock on model_one. `bypass` eventually kills the blocking lock.

Then the migration moves to the second operation, which also might encounter blocking locks. So we need to keep running bypass until the end of the wrapped code.

Comment (wesleykendall) [reply pglock/core.py]: > Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, bypass is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

> This also makes me think about the "reverse" behaviour of bypass. As I mentioned in the pgmigrate review, we often get the in prod scenario where the blocks coming from the migration after it has acquired the lock are just as damaging as the blocks preventing it acquiring the lock. Could bypass have an option that once the lock is acquired, it then starts killing PIDs which it itself is blocking?

> Alternatively, if we knew these blocking queries came from certain functions, we could decorate them and tell them not to wait for certain other functions and/or locks. Rough ideas:

```
@pglock.terminate_on('migrate')     # could take a list of function paths  or django commands for example
def my_slow_update_func():
       ...
```

OR

```
@pglock.terminate_on(pglock.ACCESS_EXCLUSIVE) 
def my_slow_update_func():
       ...
```

> The first one I'm not exactly sure how we'd do, but it's the most appealing. The second one above would already be good for avoiding clashes with migrations I think.

For the first one, I think you can better accomplish this with advisory locks. I.e take out a "shared" advisory lock on "migrate". When the migrate command runs, it will take out the "exclusive" advisory lock. If it can't take the exclusive lock, it can have a background worker that kills anything that has the shared lock. I.e. the exclusive lock always gets priority.

I.e. you could add a `prioritize` argument to `pglock.advisory` that would accomplish this.

I think it's possible to do the "reverse" behavior like you mentioned. It's just a matter of querying the `blocking_activity_id` field of `BlockedPGLock`. It is a more expensive query though since Postgres doesn't have the inverse of the `pg_blocking_pids` function

File: pglock/models.py (hunk 15 above)
Comment (wesleykendall) [reply pglock/models.py:51]: I don't know too much about it, but I know it's for storing large objects with variable sizes that span multiple pages - https://www.postgresql.org/docs/current/storage-toast.html

File: README.rst (hunk 4 above)
Comment (wesleykendall) [reply README.rst]: Thanks!

**Comment by wesleykendall** [comment]
@paul, thanks for all of the wonderful comments!

Here's what I'm planning to change after the review:

1. Remove the `wait` argument to `pglock.model` and `pglock.advisory`. The `timeout` argument should control this behavior. In order to make the code less clumsy, I think `timeout` should be able to take a floating-point number in seconds. This is the standard for functions like `time.sleep()`.

When I change the documentation, I will only document using it this way. In other words, documentation examples will look like this:

```
# Set the timeout
with pglock.timeout(1.5):
    ...

# Set the timeout on an advisory lock
with pglock.advisory(timeout=1.5):
   ...
```

I'd still like to keep the option of using a timedelta though for precision. I can just add a tip to the docs for this usage:

```
with pglock.timeout(timedelta(milliseconds=100)):
   ...
```

OR

```
with pglock.timeout(seconds=1, milliseconds=100):
   ...
```

2. I will add a `pglock.advisory_id` that will retrieve the ID used for the advisory lock. People can use this function for more advanced use cases.

In the future, I think it would be good to just bake this into the `PGLock` model/queryset and make a special option for querying by advisory locks.

3. As mentioned in the responses, there are places in which I simply need to use savepoints for correctness if timeouts are used. Hope the example makes sense.

4. I'm going to remove the ``retries`` argument to ``bypass``. It's a premature optimization that tries to solve a rare edge case. The code still functionally works without it, and we can revisit it later if it needs to be addressed.

5. After you mentioned implementing the "reverse" of `bypass`, it made me think we should just name the function `prioritize`. I think this is a better name because the function is prioritizing the locks of the wrapped code over the locks of other code. If we want to implement a "reverse" version, we can name it `deprioritize`. The implementation would be very straightforward too. I'd like to follow up later on this because there might be performance implications of implementing this that I can't test out at the moment.

6. I wasn't planning on implementing the `allow_nesting` argument to `timeout` or the ability to retrieve the return value when using `advisory` as a decorator. I think we can consider these in a later release though.

File: README.rst (hunk 2 above)
Comment (PaulGilmartin) [reply README.rst:66]: I suppose `User.objects.all().select_for_update()` would have the same effect as locking the whole table? But probably does something different under the hood.

File: docs/timeout.rst (hunk 12 above)
Comment (PaulGilmartin) [reply docs/timeout.rst:35]: I would say that's exactly the behaviour one would want from the `allow_nested=False` part, but no need to do it just now anyway.

File: pglock/core.py (hunk 13 above)
Comment (PaulGilmartin) [reply pglock/core.py:110]: Sorry I wasn't clear here. I meant the fact that you use dynamically set and check the attribute with name "value" on the local thread. It could well be that an application dev does exactly the same thing in their code but for a completely different purpose. In that scenario, `_timeout` may have the attribute "value" but it may be something else completely.

File: docs/bypass.rst (hunk 10 above)
Comment (PaulGilmartin) [reply docs/bypass.rst]: Yeah I think I'm on the same page and understand what retries does. It was just more if I was a dev using it for the behaviour you describe (i.e. wanting to not wait for the period to finish before killing more PIDs), I'd never know what finite number to choose. It's obvious in your example because we laid out the queries before hand, but of course that wouldn't really ever be a real world scenario. That's why I suggested just having a 'continual' option which just kills constantly, which is what I'd want most often in an emergency.

File: pglock/core.py (hunk 9 above)
Comment (PaulGilmartin) [reply pglock/core.py:236]: Sounds good

File: pglock/core.py (hunk 8 above)
Comment (PaulGilmartin) [reply pglock/core.py:288]: Maybe I'm just being stupid, but I think I'd find that a little confusing as an end user. I think I would/expect want the whole transaction to fail in the scenario you give (especially since advisory locks do not ever take out a transaction in postgres, I wouldn't expect nesting an advisory here to behave like nesting an atomic).

File: README.rst (hunk 2 above)
Comment (wesleykendall) [reply README.rst:66]: Not if you want to block inserting new rows or selecting from the table. I'll see if I can find some better examples we can eventually put in the docs

File: pglock/core.py (hunk 8 above)
Comment (wesleykendall) [reply pglock/core.py:288]: I think there might be some confusion here. Note that I only
do the savepoint if already in a transaction. If not in a transaction, a new one won't be started.
The savepoint *only* surrounds the call to the `pg_advisory..` function when a timeout is set. It doesn't surround
the entire wrapped block.

Consider this:

```
with transaction.atomic():
    with pglock.advisory(timeout=1) as acquired:
        if not acquired:
            # Query some information if the advisory lock cannot be acquired
            users = User.objects.all()     
        else:
            # Do another operation if you have the lock
```

You're saying that you expect line 5 to throw an exception if the lock can't be acquired? That's what it will
do without the savepoint.

Consider this - let's make a function that grabs an advisory lock:

```
def my_function():
    with pglock.advisory(timeout=0) as acquired:
        if not acquired:
            return False
        else:
            # Do some atomic action
            return True
```

Now call that function in a transaction. If ``False`` is returned, every query will
fail afterwards.

```
@transaction.atomic()
def my_other_function():
    val = my_function()

    # Every query will fail if my_function returns False
```
